src/ferny/ssh_askpass.py

```diff
@@ -0,0 +1,181 @@
+import logging
+import re
+from typing import ClassVar, Dict, Optional, Sequence
+
+from .interaction_agent import AskpassHandler
+
+logger = logging.getLogger(__name__)
+
+
+class AskpassPrompt:
+    stderr: str
+    messages: str
+    prompt: str
+
+    def reply(self, response: str) -> None:
+        pass
+
+    def close(self) -> None:
+        pass
+
+    async def handle_via(self, responder: 'SshAskpassResponder') -> None:
+        try:
+            response = await self.dispatch(responder)
+            if response is not None:
+                self.reply(response)
+        finally:
+            self.close()
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_prompt(self)
+
+
+class SSHAskpassPrompt(AskpassPrompt):
+    ANSWERS: ClassVar[Optional[Sequence[str]]] = None
+    EXTRA_PATTERNS: ClassVar[Sequence[str]] = ()
+    PATTERN: ClassVar[str]
+
+    def __init__(self, matches: Optional[Dict[str, str]] = None) -> None:
+        if matches is not None:
+            self.__dict__.update(matches)
+
+
+# Specific prompts
+HELPERS = {
+    "%{algorithm}": r"(?P<algorithm>\b[-\w]*\b)",
```

Comment:
+ perhaps? * would match on a totally empty string. The `\b` probably save us here, but still feels brittle.

Reply:
I guess you're advocating for `+`?

Reply:
Yes, I was -- but GitHub formats it as itemization bullet :see_no_evil:

Reply:
done.